### PR TITLE
[dynamo][guards-c++-refactor] EQUALS_MATCH guard

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120123
* #120119
* #120098
* #120096
* #120093
* #120091
* #120089
* #120068
* #120067
* #120065
* #120064
* #120062
* #120061
* #120060
* #119833
* __->__ #119827
* #119822

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -822,6 +822,29 @@ class ID_MATCH : public LeafGuard {
...
 
+class EQUALS_MATCH : public LeafGuard {
+ public:
+  EQUALS_MATCH(py::object value, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts),
+        _value(value),
+        _value_type(Py_TYPE(value.ptr())) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return Py_TYPE(value) == _value_type &&
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:833]: I think some of the existing usage of this compare different types -- for example we compare torch.Size to tuple.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:833]: There were 3 special cases here

1,2) numpy.isnan/math.isnan - Handled separately in the prototype [here](https://github.com/pytorch/pytorch/pull/110590/files#diff-61187d2fc8c5734622bcf785b2f864dca420e5033f9c152e413a4bcd8c143a5fR669-R671) by a separate LAMBDA_GUARD
3) torch.Size - Handled separately in the prototype [here](https://github.com/pytorch/pytorch/pull/110590/files#diff-9f0663783bcd93e948e0491ef61b48123bdc9977bcc632fd707da578df13bfa1R792)  by adding a TYPE_MATCH guard.

So, I decided to keep this one simple and rely on an additional guard for simplicity.

**Review by jansel — APPROVED** [review]

File: torch/csrc/dynamo/guards.cpp [hunk 2]
```diff
@@ -836,6 +836,29 @@ class ID_MATCH : public LeafGuard {
...
+        PyObject_RichCompareBool(value, _value.ptr(), Py_EQ);
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:848]: Need to check for NULL return, add a fast path for ptr equals.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp (hunk 2 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp]: Done. Ready for re-review.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**